### PR TITLE
Resharding: leader rotation.

### PR DESCRIPTION
**Pull request by Frozen**
Rotate leader. Currently we choose one leader for all epoch lifetime and change only if leader goes down. PR changes this, so new leader changes every 64 blocks. It calculates blocks count produced by leader, and move to next leader when  enough blocks. It doesn't conflict with current view change behaviour.

### CODE REVIEW

File: consensus/consensus_service.go
Comment (rlan35) [inline consensus/consensus_service.go:84]: why do such change? and deviate from simply getting the first key in the committee list and the first leader?

File: consensus/consensus_service.go
Comment (rlan35) [inline consensus/consensus_service.go:102]: when the committee (publicKeys) are updated, that means the epoch is switching, we should just default back to the first key as the leader. Why call updateLeader() here?

File: consensus/consensus_v2.go
Comment (rlan35) [inline consensus/consensus_v2.go:742]: why do this change? why not directly fetch the epoch from the proposed block? that should be already correct.

File: consensus/consensus_service.go
Comment (Frozen) [reply consensus/consensus_service.go]: rolled back

File: consensus/consensus_v2.go
Comment (rlan35) [inline consensus/consensus_v2.go:700]: We should keep a local cache of the leader situation for past blocks, instead of reading from the DB and parse the coinbase for every block it's pretty heavy. This DB reading and parsing should only happen when there is not enough cache data in memory to use (e.g. node first startup and for the first N blocks)

File: consensus/consensus_v2.go
Comment (rlan35) [inline consensus/consensus_v2.go:716]: Let's remove the allowlist logic for now. We may not need to use it eventually.

File: consensus/consensus_v2.go
Comment (Frozen) [reply consensus/consensus_v2.go]: cached

File: consensus/consensus_v2.go
Comment (Frozen) [reply consensus/consensus_v2.go]: Removed. I can remove `IsAllowlistEpoch` method and logic if we don't need it anymore in another pr.

File: internal/params/config.go
Comment (MaxMustermann2) [inline internal/params/config.go:351]: In addition to this `TODO`, please also add `LeaderRotationEpoch` to be `EpochTBD` for all non-testnet chains.

File: consensus/consensus_v2.go
Comment (MaxMustermann2) [inline consensus/consensus_v2.go:687]: Why are we checking this again?

File: consensus/consensus_v2.go
Comment (MaxMustermann2) [inline consensus/consensus_v2.go:694]: Nit: Previous `block was epoch block`, we should not change leader.

File: node/api.go
Comment (MaxMustermann2) [inline node/api.go:23]: Where is this used?

File: node/api.go
Comment (MaxMustermann2) [inline node/api.go:24]: Why not `node.Consensus.GetPublicKeys()` ?

File: internal/utils/singleton.go
Comment (Frozen) [reply internal/utils/singleton.go:200]: It's useful for debugging. I prefer keep it, it's used often.

File: consensus/consensus_service.go
Comment (Frozen) [reply consensus/consensus_service.go:451]: No use currently, but better keep it, it will be overwritten in next PR.

File: consensus/consensus_v2.go
Comment (Frozen) [reply consensus/consensus_v2.go:687]: This check is cheap, we can keep it. Previously this function was called from different places, so we needed it.

File: node/api.go
Comment (Frozen) [reply node/api.go:23]: This function will be overwritten in next PR https://github.com/harmony-one/harmony/pull/4300/files#diff-b1f5bb00d842600c8142ae84ffba5888cbcc451590063e43198771d4d0b6a1b4R167 . We can keep it now, i'll change it in next pr.

File: consensus/consensus_v2.go
Comment (Frozen) [reply consensus/consensus_v2.go:694]: fixed

File: internal/params/config.go
Comment (Frozen) [reply internal/params/config.go:351]: fixed

File: internal/utils/singleton.go
Comment (MaxMustermann2) [reply internal/utils/singleton.go:200]: Agreed; it's very practical. But please add a comment indicating its unused in the code and only used during development.

Comment (Frozen) [reply internal/utils/singleton.go:201]: Added comment.

File: consensus/consensus_service.go
Comment (Frozen) [reply consensus/consensus_service.go:451]: Actually, this function does what described. I don't see any problems here.

**Review by ONECasey — CHANGES_REQUESTED** [review]
@Frozen , can you please fix the conflicts?

**Comment by ONECasey** [comment]
Fixed conflicts.

**Review by ONECasey — APPROVED** [review]
Approved. Looks good to go into `dev` once state sync has been merged into `main`

**Pull request merged**
